controllers/space/space_controller_test.go

```diff
@@ -1029,6 +1031,165 @@ func TestUpdateSpaceTier(t *testing.T) {
 	})
 }
 
+// TestUpdateSpaceRoles covers the cases where SpaceBindings are created/updated/deleted,
+// and how this should affect the Space and its NSTemplateSet
+func TestUpdateSpaceRoles(t *testing.T) {
+
+	// given
+	logf.SetLogger(zap.New(zap.UseDevMode(true)))
+	s := scheme.Scheme
+	err := apis.AddToScheme(s)
+	require.NoError(t, err)
+	basicTier := tiertest.BasicTier(t, tiertest.CurrentBasicTemplates)
+
+	t.Run("add user with admin role", func(t *testing.T) {
+		// given a MUR, a Space and its NSTemplateSet resource...
+		adminMUR := murtest.NewMasterUserRecord(t, "jack")
+		viewerMUR := murtest.NewMasterUserRecord(t, "jeff")
+		s := spacetest.NewSpace("oddity",
+			spacetest.WithTierName(basicTier.Name),
+			spacetest.WithSpecTargetCluster("member-1"),
+			spacetest.WithStatusTargetCluster("member-1"), // already provisioned on a target cluster
+			spacetest.WithFinalizer())
+		nstmplSet := nstemplatetsettest.NewNSTemplateSet("oddity",
+			nstemplatetsettest.WithReferencesFor(basicTier,
+				// include pre-existing users with role...
+				nstemplatetsettest.WithSpaceRole("admin", adminMUR.Name),
+				nstemplatetsettest.WithSpaceRole("viewer", viewerMUR.Name),
+			),
+			nstemplatetsettest.WithReadyCondition(),
+		)
+		// ...and their corresponding space bindings
+		sb1 := spacebindingtest.NewSpaceBinding(adminMUR.Name, s.Name, "admin", "signupAdmin")
+		sb2 := spacebindingtest.NewSpaceBinding(viewerMUR.Name, s.Name, "viewer", "signupViewer")
+
+		// and a SpaceBinding for John as an Admin on the Space
+		johnMUR := murtest.NewMasterUserRecord(t, "john")
+		sb3 := spacebindingtest.NewSpaceBinding(johnMUR.Name, s.Name, "admin", "signupJohn")
+
+		hostClient := test.NewFakeClient(t, s, johnMUR, adminMUR, sb1, viewerMUR, sb2, sb3, basicTier)
+		member1Client := test.NewFakeClient(t, nstmplSet)
+		member1 := NewMemberClusterWithClient(member1Client, "member-1", corev1.ConditionTrue)
+		member2 := NewMemberCluster(t, "member-2", corev1.ConditionTrue)
+
+		ctrl := newReconciler(hostClient, member1, member2)
+
+		// when
+		res, err := ctrl.Reconcile(context.TODO(), requestFor(s))
+
+		// then
+		require.NoError(t, err)
+		assert.True(t, res.Requeue) // expect a requeue since the NSTemplateSet was updated
+		// Space should be in "updating" state while the
+		spacetest.AssertThatSpace(t, test.HostOperatorNs, s.Name, hostClient).
+			HasConditions(spacetest.Updating())
+		// NSTemplateSet should have an spaceRoles entry for the `mur`
+		nstemplatetsettest.AssertThatNSTemplateSet(t, test.MemberOperatorNs, nstmplSet.Name, member1Client).
+			HasSpaceRoleForUser("basic-admin-123456new", johnMUR.Name).    // entry added for user as admin
+			HasNoSpaceRoleForUser("basic-viewer-123456new", johnMUR.Name). // but not added as viewer
+			HasSpaceRoleForUser("basic-admin-123456new", adminMUR.Name).   // unchanged
+			HasSpaceRoleForUser("basic-viewer-123456new", viewerMUR.Name). // unchanged
```

Comment:
Personally, I would rather use something like this:

```go
nstemplatetsettest.AssertThatNSTemplateSet(t, test.MemberOperatorNs, nstmplSet.Name, member1Client).
	HasSpaceRoles(
		SpaceRole("basic-admin-123456new", johnMUR.Name, adminMUR.Name),
		SpaceRole("basic-viewer-123456new", viewerMUR.Name)).
```
In that way, the `HasSpaceRoles` method could easily verify that there is exactly this set of SpaceRoles defined (not more) and at the same time the `SpaceRole` function would verify that it has the exact set of usernames (not more either). So, compared to your version it would verify also the exact count/set of the entries plus it would be shorter. 
Just to be clear, this is a proposal - not a request that would block a PR ;-).

Reply:
good point, let me do that

Reply:
done in de6d17a